zerver/webhooks/newrelic/view.py

```diff
@@ -1,49 +1,78 @@
-from zerver.lib.validator import check_dict
+from zerver.lib.response import json_error, json_success
 from zerver.lib.webhooks.common import check_send_webhook_message
 from zerver.models import UserProfile
 
-ALERT_TEMPLATE = "{long_description} ([view alert]({alert_url}))."
-
-DEPLOY_TEMPLATE = """
-**{revision}** deployed by **{deployed_by}**:
-
+OPEN_TEMPLATE = """
+Incident **opened** for condition: **{condition_name}** at <time:{iso_timestamp}>
 ``` quote
-{description}
+{details}
 ```
+[View incident]({incident_url})
+""".strip()
 
-Changelog:
+DEFAULT_TEMPLATE = """Incident **{status}** for condition: **{condition_name}**""".strip()
 
-``` quote
-{changelog}
-```
-""".strip()
+TOPIC_TEMPLATE = """{policy_name} ({incident_id})""".strip()
 
 @webhook_view("NewRelic")
 @has_request_variables
-def api_newrelic_webhook(request: HttpRequest, user_profile: UserProfile,
-                         alert: Optional[Dict[str, Any]]=REQ(validator=check_dict([]), default=None),
-                         deployment: Optional[Dict[str, Any]]=REQ(validator=check_dict([]), default=None),
-                         ) -> HttpResponse:
-    if alert:
-        # Use the message as the subject because it stays the same for
-        # "opened", "acknowledged", and "closed" messages that should be
-        # grouped.
-        subject = alert['message']
-        content = ALERT_TEMPLATE.format(**alert)
-    elif deployment:
-        subject = "{} deploy".format(deployment['application_name'])
-        content = DEPLOY_TEMPLATE.format(**deployment)
+def api_newrelic_webhook(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    payload: Dict[str, Any]=REQ(argument_type='body')
+) -> HttpResponse:
+
+    info = {
+        "condition_name": payload.get('condition_name', 'Unknown condition'),
+        "details": payload.get('details', 'No details.'),
+        "incident_url": payload.get('incident_url', 'https://alerts.newrelic.com'),
+        "incident_acknowledge_url": payload.get('incident_acknowledge_url', 'https://alerts.newrelic.com'),
+        "status": payload.get('current_state', 'None'),
+        "iso_timestamp": '',
+    }
+
+    unix_time = payload.get('timestamp', None)
+    if unix_time is None:
+        return json_error(_("The newrelic webhook requires timestamp in milliseconds"))
+
+    duration = payload.get('duration', None)
+    if duration is None:
+        return json_error(_("The newrelic webhook requires duration in milliseconds"))
+
+    try:
+        # Timestamp - duration to get time alert started
+        # timestamps from NewRelic are in ms
+        iso_timestamp = datetime.fromtimestamp((unix_time - duration) / 1000)
+        info['iso_timestamp'] = iso_timestamp
+    except ValueError:
+        return json_error(_("The newrelic webhook expects timestamp and duration in milliseconds"))
+    except TypeError:
+        return json_error(_("The newrelic webhook expects timestamp and duration in milliseconds"))
```

Comment:
Can we move the time-formatting logic for generating `<time` syntax to common code in `zerver/lib/webhooks/common.py`?  We'd need to make the exception use the name of the integration and JsonableError, but that'd make it easier to reuse for other webhooks.

Reply:
I am not quite sure I understand this one. 
Would I add a subclass of JsonableError and then a unix_milleseconds_to_timestamp method, or would I just add the unix_milleseconds_to_timestamp method that takes in milliseconds and integration_name and throw a JsonableError?

Reply:
I merged without addressing this since it's a fine follow-up PR.  The latter is the API I had in mind, and I think that's cleaner; there's no need for a special subclass that'll only be used in one place.